src/shared/profiling.ts

```diff
@@ -0,0 +1,95 @@
+/*!
+ * Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { getLogger } from './logger/logger'
+
+const performance = globalThis.performance ?? require('perf_hooks').performance
+const times: {
+    [name: string]: {
+        time: number
+        parent: string | undefined
+        children: string[]
+    }
+} = {}
+
+let currentScope = 'toplevel'
+
+/**
+ * Executes function `fn` with the given arguments and records the elapsed time.
+ *
+ * TODO: merge this into TelemetryTracer.run() or a new TelemetryTracer.runTimed() function.
+ *
+ */
+export function timed<T>(name_: string | undefined, fn: (...args: any[]) => Promise<T>, ...args: any[]): Promise<T>
+export function timed<T>(name_: string | undefined, fn: (...args: any[]) => T, ...args: any[]): T
+export function timed<T>(
+    name_: string | undefined,
+    fn: ((...args: any[]) => T) | ((...args: any[]) => Promise<T>),
+    ...args: any[]
+): T | Promise<T> {
+    const name = name_ ? name_ : fn.name
+    if (performance.getEntriesByName(name).length) {
+        throw Error(`name must be non-empty and unique: "${name}"`)
+        // throw Error('name must be non-empty')
+    }
+    times[name] = times[name] ?? { time: 0, parent: undefined, children: [] }
+    if (times[currentScope]) {
+        times[name].parent = currentScope
+        times[currentScope].children.push(name)
+    }
+    const oldScope = currentScope
+    currentScope = name
+    const start = performance.now()
+    // performance.mark(name, { startTime: start })
+
+    const after = () => {
+        const m = performance.measure(name, { start: start, end: performance.now() })
+        times[name].time = m.duration
+        currentScope = oldScope // TODO: deal with concurrent async executions.
```

Comment:
Oh I was working on a "task" module that does this. You could hook into it to do perf analysis.